File: ios/App/Snappet/Features/Kilter/KilterDuplicateChecker.swift
import Foundation

/// Validates a freshly-authored climb (manual *or* generated) against everything already on the device
/// before it's saved — the user's brief: *"when a user saves, the app should validate that climb with all
/// the downloaded dataset for duplicates; same applies for model-generated climbs."*
///
/// Duplicate ≡ same **canonical** hold set on the same layout (`KilterClimbIdentity.canonicalKey`), so the
/// match is order-independent: placing the same holds in a different sequence still collides. The index
/// covers both sources:
/// - the read-only **catalog** (the downloaded dataset) — matched by canonical frames, since catalog
///   climbs carry Aurora's own random uuids, not our content uuid;
/// - the user's previously **created** climbs — which *do* carry the deterministic content uuid, so the
///   same climb authored twice (even on another device, then synced in) collapses onto one entry.
///
/// Layout-scoped (built per layout) because a `placement_id` only means something within its layout. Pure
/// core (`init` + `find`) is unit-tested without a device; `build(...)` is the main-actor convenience that
/// reads the installed catalog.
struct KilterDuplicateChecker {
    /// A climb the authored one duplicates.
    struct Duplicate: Equatable {
        enum Origin: Equatable {
            case catalog   // an existing climb in the downloaded dataset
            case created   // one the user already authored on this device
        }
        let uuid: String
        let name: String
        let setter: String
        let origin: Origin
    }

    /// `canonicalKey -> Duplicate`.
    private let index: [String: Duplicate]
    let layoutId: Int

    /// Build the index from raw catalog rows + the user's created climbs (both already scoped to one
    /// layout). A created climb shadows a catalog climb with the same holds (the user's own copy wins in
    /// the "already exists" message). Pure → testable with hand-built rows.
    init(catalogRows: [(uuid: String, name: String, setter: String, frames: String)],
         createdClimbs: [(uuid: String, name: String, setter: String, layoutId: Int, frames: String)],
         layoutId: Int) {
        self.layoutId = layoutId
        var idx: [String: Duplicate] = [:]
        for r in catalogRows {
            let key = KilterClimbIdentity.canonicalKey(layoutId: layoutId, frames: r.frames)
            if idx[key] == nil {
                idx[key] = Duplicate(uuid: r.uuid, name: r.name, setter: r.setter, origin: .catalog)
            }
        }
        for c in createdClimbs where c.layoutId == layoutId {
            let key = KilterClimbIdentity.canonicalKey(layoutId: layoutId, frames: c.frames)
            idx[key] = Duplicate(uuid: c.uuid, name: c.name, setter: c.setter, origin: .created)
        }
        self.index = idx
    }

    /// The existing climb this `(layout, frames)` duplicates, or `nil` if it's new. The frames need not be
    /// canonical — `find` canonicalizes before lookup.
    func find(layoutId: Int, frames: String) -> Duplicate? {
        guard layoutId == self.layoutId else { return nil }
        return index[KilterClimbIdentity.canonicalKey(layoutId: layoutId, frames: frames)]
    }

    /// Build a checker for one layout from the installed catalog + the user's created climbs. Main-actor
    /// because it reads `KilterCatalog` (main-thread-only). Reads the catalog's climbs once per call —
    /// cheap (one indexed SELECT over a layout) and always current, so it's built fresh at save time.
    @MainActor
    static func build(layoutId: Int, created: [KilterCreatedClimb],
                      catalog: KilterCatalog = .shared) -> KilterDuplicateChecker {
        let rows = catalog.climbFramesForDedup(forLayout: layoutId)
        let createdTuples = created.map {
            (uuid: $0.uuid, name: $0.name, setter: $0.setterUsername, layoutId: $0.layoutId, frames: $0.frames)
        }
        return KilterDuplicateChecker(catalogRows: rows, createdClimbs: createdTuples, layoutId: layoutId)
    }
}

File: android/app/src/main/java/com/snappet/mobile/feature/kilter/KilterDuplicateChecker.kt
package com.snappet.mobile.feature.kilter

/**
 * Validates a freshly-authored climb (manual *or* generated) against everything already on the device
 * before it's saved — same brief as iOS: dedupe against the downloaded dataset *and* prior creations.
 *
 * Duplicate ≡ same **canonical** hold set on the same layout ([KilterClimbIdentity.canonicalKey]), so the
 * match is order-independent. The index covers the read-only catalog (matched by canonical frames, since
 * catalog climbs carry Aurora's random uuids) and the user's created climbs (which carry the deterministic
 * content uuid). Layout-scoped. Pure core ([find]) is unit-tested; [build] reads the installed catalog.
 * Mirrors iOS `KilterDuplicateChecker`.
 */
class KilterDuplicateChecker(
    catalogRows: List<CatalogRow>,
    createdClimbs: List<CreatedRow>,
    private val layoutId: Int,
) {
    /** Where a duplicate came from. */
    enum class Origin { CATALOG, CREATED }

    data class Duplicate(val uuid: String, val name: String, val setter: String, val origin: Origin)
    data class CatalogRow(val uuid: String, val name: String, val setter: String, val frames: String)
    data class CreatedRow(val uuid: String, val name: String, val setter: String, val layoutId: Int, val frames: String)

    private val index: Map<String, Duplicate> = buildMap {
        for (r in catalogRows) {
            val key = KilterClimbIdentity.canonicalKey(layoutId, r.frames)
            if (!containsKey(key)) put(key, Duplicate(r.uuid, r.name, r.setter, Origin.CATALOG))
        }
        for (c in createdClimbs) if (c.layoutId == layoutId) {
            val key = KilterClimbIdentity.canonicalKey(layoutId, c.frames)
            put(key, Duplicate(c.uuid, c.name, c.setter, Origin.CREATED))   // created shadows catalog
        }
    }

    /** The existing climb this `(layout, frames)` duplicates, or null if new. Frames need not be canonical. */
    fun find(layoutId: Int, frames: String): Duplicate? {
        if (layoutId != this.layoutId) return null
        return index[KilterClimbIdentity.canonicalKey(layoutId, frames)]
    }

    companion object {
        /** Build a checker for one layout from the installed catalog + the user's created climbs. */
        fun build(layoutId: Int, catalog: KilterCatalog, created: List<KilterCreatedClimb>): KilterDuplicateChecker {
            val rows = catalog.climbFramesForDedup(layoutId)
            val createdRows = created.map {
                CreatedRow(it.uuid, it.name, it.setterUsername, it.layoutId, it.frames)
            }
            return KilterDuplicateChecker(rows, createdRows, layoutId)
        }
    }
}
